### PR TITLE
send webhook when task or workflow run is canceled

### PR DESCRIPTION
**Pull request by wintonzheng**
> [!IMPORTANT]
> Send webhooks when tasks or workflow runs are canceled by updating `execute_step()`, `execute_workflow()`, and related functions to handle webhook execution.
> 
>   - **Behavior**:
>     - Send webhook when task or workflow run is canceled by setting `need_call_webhook=True` in `execute_step()` in `agent.py` and `execute_workflow()` in `service.py`.
>     - Modify `cancel_task()` and `cancel_workflow_run()` in `agent_protocol.py` to include `x_api_key` for webhook execution.
>   - **Functions**:
>     - Update `execute_task_webhook()` in `agent.py` to accept `last_step` as `Step | None`.
>     - Add `execute_workflow_webhook()` in `service.py` to handle webhook execution for workflow runs.
>   - **Misc**:
>     - Minor logging adjustments in `service.py` for webhook execution.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=Skyvern-AI%2Fskyvern&utm_source=github&utm_medium=referral)<sup> for ac4a400e563063a02d3253c5d70f0d153fec6c5f. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: skyvern/forge/sdk/routes/agent_protocol.py [hunk 1]
```diff
@@ -292,16 +293,28 @@ async def cancel_task(
             status_code=status.HTTP_404_NOT_FOUND,
             detail=f"Task not found {task_id}",
         )
-    await app.agent.update_task(task_obj, status=TaskStatus.canceled)
+    task = await app.agent.update_task(task_obj, status=TaskStatus.canceled)
```
Comment (LawyZheng) [inline skyvern/forge/sdk/routes/agent_protocol.py:296]: shall we just only send the webhook when task/workflow cancellation is finished? any specific reason why we must send the webhook in the API call?

File: skyvern/forge/sdk/routes/agent_protocol.py [hunk 2]
```diff
@@ -292,16 +293,28 @@ async def cancel_task(
...
+    x_api_key: Annotated[str | None, Header()] = None,
 ) -> None:
+    workflow_run = await app.DATABASE.get_workflow_run(workflow_run_id=workflow_run_id)
+    if not workflow_run:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail=f"Workflow run not found {workflow_run_id}",
+        )
     await app.WORKFLOW_SERVICE.mark_workflow_run_as_canceled(workflow_run_id)
+    await app.WORKFLOW_SERVICE.execute_workflow_webhook(workflow_run, api_key=x_api_key)
```
Comment (LawyZheng) [inline skyvern/forge/sdk/routes/agent_protocol.py:317]: same question as task

File: skyvern/forge/sdk/routes/agent_protocol.py (hunk 1 above)
Comment (wintonzheng) [reply skyvern/forge/sdk/routes/agent_protocol.py:296]: users can also cancel tasks that are stuck. our worker won't really send a webhook if we don't send a webhook to them in the API endpoint.

Comment (wintonzheng) [reply skyvern/forge/sdk/routes/agent_protocol.py:296]: i think it's okay that we send two cancel task webhook.

**Pull request merged**